### PR TITLE
raft: cleanup read index code

### PR DESCRIPTION
**Pull request by hicqu**
Signed-off-by: qupeng <qupeng@pingcap.com>

Cleanup code about read index, extracted from #11505 .

### CODE REVIEW

**Comment by hicqu** [comment]
@xiang90 PTAL if you are free, thanks a lot!

File: raft/raft.go [hunk 1]
```diff
@@ -1602,6 +1594,25 @@ func (r *raft) abortLeaderTransfer() {
 	r.leadTransferee = None
 }
 
+func (r *raft) CommittedEntryInCurrentTerm() bool {
```
Comment (xiang90) [inline raft/raft.go]: does this func need to be public?

File: raft/raft.go [hunk 2]
```diff
@@ -1602,6 +1594,25 @@ func (r *raft) abortLeaderTransfer() {
...
+			Index:      committed,
+			RequestCtx: req.Entries[0].Data,
+		})
+		internal = true
+	}
+	resp.Type = pb.MsgReadIndexResp
+	resp.To = req.From
+	resp.Index = committed
+	resp.Entries = req.Entries
+	return
```
Comment (xiang90) [inline raft/raft.go]: kill naked return.

File: raft/raft.go [hunk 3]
```diff
@@ -1602,6 +1594,25 @@ func (r *raft) abortLeaderTransfer() {
 	r.leadTransferee = None
 }
 
+func (r *raft) CommittedEntryInCurrentTerm() bool {
+	return r.raftLog.zeroTermOnErrCompacted(r.raftLog.term(r.raftLog.committed)) == r.Term
+}
+
+func (r *raft) responseReadIndex(req pb.Message, committed uint64) (resp pb.Message, internal bool) {
```
Comment (xiang90) [inline raft/raft.go]: rename to ResponseToReadIndexReq?

1. why do you need to pass in the committed index? you can get it from the method receiver itself?

2. what does internal mean in the return?

Comment (hicqu) [reply raft/raft.go]: > why need to pass in the committed index
For leader, the index should be its committed index when it receiving the `ReadIndex` request, not its committed index after it gets enough heartbeat responses. So we must pass it into the function.

**Comment by hicqu** [comment]
@xiang90 CI seems fail because of other reasons, not releated with this PR. Could you help me to re-run CI?

File: raft/raft.go (hunk 1 above)
Comment (xiang90) [reply raft/raft.go]: @hicqu this needs to be fixed.

File: raft/raft.go [hunk 4]
```diff
@@ -1602,6 +1594,29 @@ func (r *raft) abortLeaderTransfer() {
...
 }
 
+// CommittedEntryInCurrentTerm return true if the peer has committed an entry in its term.
+func (r *raft) CommittedEntryInCurrentTerm() bool {
+	return r.raftLog.zeroTermOnErrCompacted(r.raftLog.term(r.raftLog.committed)) == r.Term
+}
+
+// responseToReadIndexReq constructs a response for `req`. If `req` comes from the peer
+// itself, a blank value will be returned.
+func (r *raft) responseToReadIndexReq(req pb.Message, committed uint64) pb.Message {
```
Comment (xiang90) [inline raft/raft.go]: it should not be `committed`. it is the safe index that the requested node can read. if it is the local node, then then current commit index is OK. if it is from a remote node, the original read index should be preserved.  

probably rename `committed` to `readIndex`

**Comment by hicqu** [comment]
Sorry @xiang90 I forgot to update the PR. Now I have addressed your last comment. PTAL if you are free, thanks!

File: raft/raft.go (hunk 1 above)
Comment (hicqu) [reply raft/raft.go]: Fixed.

**Comment by xiang90** [comment]
lgtm. Thanks @hicqu

**Comment by hicqu** [comment]
Friendly ping @xiang90 .

**Pull request merged**